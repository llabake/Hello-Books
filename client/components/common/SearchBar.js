import React, { Component } from 'react';
import { connect } from 'react-redux';

/**
 * 
 * 
 * @class SearchBar
 * @extends {Component}
 */
class SearchBar extends Component {

  /**
   * Creates an instance of SearchBar.
   * @param {any} props 
   * @memberof SearchBar
   */
  constructor(props) {
    super(props);
    this.state = {
      searchTerm: '',
    };
    
    this.handleChange = this.handleChange.bind(this)
    this.handleSearch = this.handleSearch.bind(this)
    this.redirectToSearch = this.redirectToSearch.bind(this)
  }

  /**
   * @return {object} state object
   * 
   * @param {any} event 
   * @memberof SearchBar
   */
  handleChange (event) {
    this.setState({
      [event.target.name]: event.target.value
    })
  }

  
  /**
   * @returns {array} an array of books that match the search criteria
   * 
   * @param {any} event 
   * @memberof SearchBar
   */
  handleSearch(event) {
    event.preventDefault();
    this.redirectToSearch();
    
  }

  /**
   * @returns {object} redirects to search page
   * 
   * @memberof SearchBar
   */
  redirectToSearch(){
    window.location.href = `/search/${this.state.searchTerm}`
  }


  /**
   * 
   * 
   * @returns {void} 
   * @memberof SearchBar
   */
  render () {
    return (
      <form onSubmit={this.handleSearch}>
        <div className="input-field col m6 s12 l6 " style={{'marginRight': '12px'}}>
          <input
            id="search" 
            type="search"  
            placeholder="Title, author, or ISBN" 
            name='searchTerm' 
            onChange={this.handleChange}
            value={this.state.searchTerm}
          />
          <label className="label-icon icon-sit" htmlFor="search"><a><i className="material-icons">search</i></a></label>
          {/* <label className="label-icon icon-sit" htmlFor="search"><button style={{
            'lineHeight': '15px',
            'letterSpacing': 'undefined',
            'paddingTop': '0px' ,
            'paddingBottom': '0px',
            'color': '#ffffff',
            'paddingLeft': '18px',
            'paddingRight': '18px',
            'borderRadius': '4px',
            'border': '0px',
            'background': '#7c4561',
            'margin': '12px ',
            'marginLeft': '16px' ,
            'height': '48px',
            'textAlign': 'center',
            'display': 'inline-block',
            'float': 'right',
            'whiteSpace': 'nowrap'
            }}>Search</button></label> */}
          <i className="material-icons">close</i>
        </div>
      </form>
    )
  }

}

const mapStateToProps = (state) => {
  return {
    errors: state.errors,
    searchedBooks: state.bookReducer.searchResult
  }
}

export default connect(mapStateToProps, null)(SearchBar);