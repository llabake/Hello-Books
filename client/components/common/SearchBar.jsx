import React, { Component } from 'react';
import { connect } from 'react-redux';

/**
 * 
 * 
 * @class SearchBar
 * @extends {Component}
 */
class SearchBar extends Component {

  /**
   * Creates an instance of SearchBar.
   * @param {any} props 
   * @memberof SearchBar
   */
  constructor(props) {
    super(props);
    this.state = {
      searchTerm: '',
    };

    this.handleChange = this.handleChange.bind(this)
    this.handleSearch = this.handleSearch.bind(this)
    this.redirectToSearch = this.redirectToSearch.bind(this)
  }

  /**
   * @return {object} state object
   * 
   * @param {any} event 
   * @memberof SearchBar
   */
  handleChange(event) {
    this.setState({
      [event.target.name]: event.target.value
    })
  }


  /**
   * @returns {array} an array of books that match the search criteria
   * 
   * @param {any} event 
   * @memberof SearchBar
   */
  handleSearch(event) {
    event.preventDefault();
    this.redirectToSearch();

  }

  /**
   * @returns {object} redirects to search page
   * 
   * @memberof SearchBar
   */
  redirectToSearch() {
    const trimmedSearchTerm = this.state.searchTerm && this.state.searchTerm.trim();
    trimmedSearchTerm ? 
    window.location.href = `/search/${trimmedSearchTerm}`:
    window.location.href = window.location.href 
  }


  /**
   * 
   * 
   * @returns {void} 
   * @memberof SearchBar
   */
  render() {
    return (
      <form onSubmit={this.handleSearch}>
        <div className="input-field col s6 s12 ">
          <input
            // id="search"
            type="search"
            placeholder="Title, Author, Keyword"
            name='searchTerm'
            onChange={this.handleChange}
            value={this.state.searchTerm}
          />
          <label
            className="label-icon icon-sit"
            htmlFor="search">
            <i
              className="material-icons clickable"
              onClick={this.handleSearch}>search
            </i>
          </label>
          <i className="material-icons"
          style={{marginTop: '-6px', fontSize: '1.5rem'}}
          >close</i>
        </div>
      </form>
    )
  }

}

const mapStateToProps = (state) => {
  return {
    errors: state.errors,
    searchedBooks: state.bookReducer.searchResult
  }
}

export default connect(mapStateToProps, null)(SearchBar);