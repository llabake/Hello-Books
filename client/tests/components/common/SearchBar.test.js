
import React from 'react';

import { SearchBar, mapStateToProps} from '../../../components/common/SearchBar.jsx';


const setup = () => {
  const props = {
    searchedBooks: [],
    handleSearch: jest.fn(),
    handleChange: jest.fn(),
  };
  const event = {
    preventDefault: jest.fn(),
    target: {
      value: 'blink',
      name: 'searchTerm'
    }
  };
  const wrapper = shallow(<SearchBar {...props} />);
  return {
    wrapper,
    props, 
    event
  };
}

test('<SearchBar />', () => {
const { wrapper, props } = setup();
  expect(wrapper).toBeDefined();
  expect(wrapper).toMatchSnapshot();
  expect(wrapper.find('form').length).toBe(1);
  expect(wrapper.find('input').length).toBe(1);
})

test('<SearchBar /> should simulate change event', () => {
  const { wrapper, props, event } = setup();
  wrapper.find('input').simulate('change', event);
  expect(wrapper.instance().state.searchTerm).toEqual('blink');
});

test('<SearchBar /> should simulate search form submission', () => {
  const event = {
    preventDefault: jest.fn(),
    target: {
      value: ''
    }
  };
  const { wrapper, props } = setup();
  const handleSearchSpy = jest.spyOn(SearchBar.prototype, 'handleSearch')
  const instance = wrapper.instance();
  const newWrapper = shallow(<SearchBar {...props}/>)
  const redirectToSearch = jest.fn()
   
  newWrapper.find('form').simulate('submit', event);
  expect(SearchBar.prototype.handleSearch).toHaveBeenCalled();
});

test('mapStateToProps', () => {
  expect(mapStateToProps({ bookReducer: {} })).toHaveProperty('searchedBooks');
});