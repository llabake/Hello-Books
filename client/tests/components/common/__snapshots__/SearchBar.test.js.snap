// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchBar /> 1`] = `
<form
  onSubmit={[Function]}
>
  <div
    className="input-field col s6 s12 "
  >
    <input
      name="searchTerm"
      onChange={[Function]}
      placeholder="Title, Author, Keyword"
      type="search"
      value=""
    />
    <label
      className="label-icon icon-sit"
      htmlFor="search"
    >
      <i
        className="material-icons clickable"
        onClick={[Function]}
      >
        search
      </i>
    </label>
    <i
      className="material-icons"
      style={
        Object {
          "fontSize": "1.5rem",
          "marginTop": "-6px",
        }
      }
    >
      close
    </i>
  </div>
</form>
`;
